File: pacotes/protocolo/src/crc16.ts
// CRC16-CCITT — polinômio 0x1021, seed 0xFFFF
// Idêntico ao algoritmo do firmware ESP32 (crc16_ccitt em main.cpp)

const TABELA_CRC16 = (() => {
  const t = new Uint16Array(256);
  for (let i = 0; i < 256; i++) {
    let crc = i << 8;
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x8000) ? ((crc << 1) ^ 0x1021) : (crc << 1);
    }
    t[i] = crc & 0xffff;
  }
  return t;
})();

/**
 * Calcula CRC16-CCITT sobre o buffer fornecido.
 * Buffer vazio retorna 0xFFFF (valor do seed inicial).
 */
export function calcularCRC16(buf: Uint8Array): number {
  let crc = 0xffff;
  for (const byte of buf) {
    crc = ((crc << 8) ^ (TABELA_CRC16[((crc >> 8) ^ byte) & 0xff] ?? 0)) & 0xffff;
  }
  return crc;
}

File: android/app/src/main/java/br/edu/ifsc/balancagfig/protocolo/Crc16.kt
package br.edu.ifsc.balancagfig.protocolo

/**
 * CRC16-CCITT — polinômio 0x1021, seed 0xFFFF.
 * Idêntico a pacotes/protocolo/src/crc16.ts e ao crc16_ccitt do firmware.
 */
object Crc16 {
    private val tabela = IntArray(256) { i ->
        var crc = i shl 8
        repeat(8) {
            crc = if (crc and 0x8000 != 0) (crc shl 1) xor 0x1021 else crc shl 1
        }
        crc and 0xffff
    }

    /** Calcula o CRC de [bytes] no intervalo [inicio, fim). Buffer vazio retorna 0xFFFF. */
    fun calcular(bytes: ByteArray, inicio: Int = 0, fim: Int = bytes.size): Int {
        var crc = 0xffff
        for (i in inicio until fim) {
            val b = bytes[i].toInt() and 0xff
            crc = ((crc shl 8) xor tabela[((crc shr 8) xor b) and 0xff]) and 0xffff
        }
        return crc
    }
}
